opencxl/apps/memory_pooling.py

```diff
@@ -6,16 +6,69 @@
 """
 
 import asyncio
+from dataclasses import dataclass
+from typing import List
 
+from opencxl.cxl.component.fabric_manager.socketio_server import HostFMMsg
+from opencxl.cxl.component.short_msg_conn import ShortMsgConn
 from opencxl.util.logger import logger
 from opencxl.drivers.cxl_bus_driver import CxlBusDriver
 from opencxl.drivers.cxl_mem_driver import CxlMemDriver
-from opencxl.drivers.pci_bus_driver import PciBusDriver
+from opencxl.drivers.pci_bus_driver import PciBusDriver, PciDeviceInfo
 from opencxl.cxl.component.cxl_memory_hub import CxlMemoryHub, MEM_ADDR_TYPE
 from opencxl.cxl.component.cxl_host import CxlHost
 from opencxl.cpu import CPU
 
 
+@dataclass
+class MemoryStruct:
+    base: int
+    size: int
+
+
+class CxlDeviceMemTracker:
+    def __init__(self, cxl_memory_hub: CxlMemoryHub):
+        self._ld_tracker: dict[str, dict[MEM_ADDR_TYPE, MemoryStruct]] = {}
+        self._cxl_memory_hub = cxl_memory_hub
+
+    def _create_key(self, device_port):
+        return f"{device_port}".lower()
+
+    def _add_device(self, key):
+        self._ld_tracker[key] = {k: MemoryStruct(0, 0) for k in MEM_ADDR_TYPE}
+
+    def check_device_added(self, device_port):
+        key = self._create_key(device_port)
+        return key in self._ld_tracker
+
+    def add_mem_range(self, device_port: int, base: int, size: int, type: MEM_ADDR_TYPE):
+        key = self._create_key(device_port)
+        if key not in self._ld_tracker:
+            self._add_device(key)
+        self._ld_tracker[key][type].base = base
+        self._ld_tracker[key][type].size = size
+        self._cxl_memory_hub.add_mem_range(base, size, type)
+
+    def remove_mem_range(self, device_port):
+        key = self._create_key(device_port)
+        if key in self._ld_tracker:
+            for type, mem_info in self._ld_tracker[key].items():
+                if mem_info.size > 0:
+                    self._cxl_memory_hub.remove_mem_range(mem_info.base, mem_info.size, type)
+            del self._ld_tracker[key]
+        else:
+            logger.warning(f"No record for device @ port {device_port}")
+
+    def print(self):
+        print(self._ld_tracker)
+
+
+@dataclass
+class MemoryBaseTracker:
+    hpa_base: int
+    cfg_base: int
```

Comment:
This tracker is only a temporary tracker for recording the current starting hpa/cfg address for new devices. For mmio, it does not need this tracker because it just reads the mmio_base from `bar_info.base_address`. I created this class because otherwise these two variables will be `int`s, so they cannot be edited in the callback closure. A class is needed to pass in these values as references.

The `CxlDeviceMemTracker` records all information about currently added hpa/cfg/mmio records.